solo-io/gloo
# Introduce GHA Cache Eviction For Closed PRs

# Description
Evict the GHA cache of a PR that has closed (this includes merged PRs)

## CI changes
- Added a new workflow to run when PRs are closed

# Context

This is documented in the GitHub docs: https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#force-deleting-cache-entries

The background is that once a PR merges, the cache associated with that branch is useless as it has been orphaned and no other PRs can utilize it. There are two options:
- Wait for the cache to be evicted (this happens when the cache grows too large, and the least recently used entry is evicted).
- Perform the eviction asap

Previously we used the first option, this PR has us use the second option.

## Testing steps

This workflow is configured to run on merge, so there isn't a great way to test this as part of the PR.

With that said, I will do the following to test this out:
1. Open the PR (should be a cache hit against main branch)
```
Cache Size: ~793 MB (831713525 B)
/usr/bin/tar -xf /home/runner/work/_temp/c29c267b-b7e6-4858-be35-63577f1b0953/cache.tzst -P -C /home/runner/work/gloo/gloo --use-compress-program unzstd
Cache restored successfully
Cache restored from key: Linux-go-1.20.7-a37322e930dd43abe0261bd0c0b1bdb0af35633393686[144](https://github.com/solo-io/gloo/actions/runs/5811217251/job/15753929151?pr=8562#step:3:151)413fa1803197fb73
```
2. Change a dependency so that a cache miss ocucrs, and there is a cache entry associated with this PR
```
Run actions/cache@v3
Cache not found for input keys: Linux-go-1.20.7-fab6aa2eed0584b187b78e2fb6f0026f7212a26d767455948288e1a05c0e5f15
```
3. Screenshot that entry
<img width="1119" alt="Screen Shot 2023-08-09 at 10 11 49 AM" src="https://github.com/solo-io/gloo/assets/5279627/c51b52f0-f3b1-434b-8df1-d9b757e66f17">

4. Update this job to run on pushes to this branch
Logs: https://pipelines.actions.githubusercontent.com/serviceHosts/9f773ffd-c435-4316-9593-73962d472ef8/_apis/pipelines/1/runs/74554/signedlogcontent/2?urlExpires=2023-08-09T16%3A21%3A14.9800890Z&urlSigningMethod=HMACV1&urlSignature=e1m8oIpdhfjKStHXOHERkVphDLaewXhxbrd5whNLQOo%3D

```
2023-08-09T16:20:08.3349588Z Fetching list of cache key
2023-08-09T16:20:08.6890695Z Deleting caches...
2023-08-09T16:20:08.6892035Z Deleting cache key Linux-go-1.20.7-fab6aa2eed0584b187b78e2fb6f0026f7212a26d767455948288e1a05c0e5f15 for branch refs/pull/8562/merge
2023-08-09T16:20:09.1354965Z [31m✓[0m Deleted 1 cache entry with key 'Linux-go-1.20.7-fab6aa2eed0584b187b78e2fb6f0026f7212a26d767455948288e1a05c0e5f15'
2023-08-09T16:20:09.1374947Z Done
```
5. Verify that the cache entry for this branch is completed
*I confirmed the cache was removed*
6. Undo all temporary changes and put this up for review

## Notes for reviewers

-

# Checklist:
